File: output/xCodeEval_8000/C++/Java/fa5bbd62e38df804504d474d91851ba6.java
import java.io.*;
import java.util.*;

class CPaintTree {
    static class Solver {
        private final InputReader in;
        private final OutputWriter out;

        Solver(InputReader in, OutputWriter out) {
            this.in = in;
            this.out = out;
        }

        int n;
        List<List<Integer>> e;
        int[] subtree_sz;
        int[] xy_mapping;
        List<Pair<Double, Integer>> xys_idx;
        List<int[]> xy;

        void readTree() {
            e = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                e.add(new ArrayList<>());
            }
            for (int i = 0; i < n - 1; i++) {
                int u = in.nextInt() - 1;
                int v = in.nextInt() - 1;
                e.get(u).add(v);
                e.get(v).add(u);
            }
        }

        void readPoints() {
            xy = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                int[] point = new int[2];
                point[0] = in.nextInt();
                point[1] = in.nextInt();
                xy.add(point);
            }
        }

        long xDist(int i, int j) {
            return xy.get(j)[0] - xy.get(i)[0];
        }

        long yDist(int i, int j) {
            return xy.get(j)[1] - xy.get(i)[1];
        }

        double getAngle(int i, int j) {
            return Math.atan2(yDist(i, j), xDist(i, j));
        }

        void sortByAngle(int l, int r, int cur_xy) {
            if (r < l) return;
            for (int i = l; i <= r; i++) {
                xys_idx.get(i).first = getAngle(cur_xy, xys_idx.get(i).second);
            }
            double maxAngle = xys_idx.get(l).first;
            for (int i = l; i <= r; i++) {
                maxAngle = Math.max(maxAngle, xys_idx.get(i).first);
            }
            for (int i = l; i <= r; i++) {
                if (xys_idx.get(i).first + Math.PI < maxAngle) {
                    xys_idx.get(i).first += 2 * Math.PI;
                }
            }
            Collections.sort(xys_idx.subList(l, r + 1));
        }

        void solveRec(int l, int r, int root, int parent) {
            int curr_xy = xys_idx.get(r--).second;
            xy_mapping[curr_xy] = root;
            sortByAngle(l, r, curr_xy);
            int cur_l = l;
            for (int i : e.get(root)) {
                if (i == parent) continue;
                int cur_r = cur_l + subtree_sz[i] - 1;
                int nxt_l = cur_r + 1;
                solveRec(cur_l, cur_r, i, root);
                cur_l = nxt_l;
            }
        }

        void calculateSubtreeSizes(int i, int p) {
            subtree_sz[i] = 1;
            for (int j : e.get(i)) {
                if (j == p) continue;
                calculateSubtreeSizes(j, i);
                subtree_sz[i] += subtree_sz[j];
            }
        }

        void solve() {
            n = in.nextInt();
            subtree_sz = new int[n];
            readTree();
            readPoints();
            xy_mapping = new int[n];
            calculateSubtreeSizes(0, -1);
            xys_idx = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                xys_idx.add(new Pair<>(0.0, i));
            }
            xys_idx.sort((i, j) -> {
                int[] pi = xy.get(i.second);
                int[] pj = xy.get(j.second);
                if (pi[0] != pj[0]) {
                    return pj[0] - pi[0];
                } else {
                    return pj[1] - pi[1];
                }
            });
            solveRec(0, n - 1, 0, -1);
            for (int i = 0; i < n; i++) {
                xy_mapping[i]++;
            }
            out.println(xy_mapping);
        }
    }

    public static void solve(InputReader in, OutputWriter out) {
        Solver solver = new Solver(in, out);
        solver.solve();
    }

    public static void main(String[] args) {
        InputReader in = new InputReader(System.in);
        OutputWriter out = new OutputWriter(System.out);
        solve(in, out);
        out.close();
    }

    static class InputReader {
        private final BufferedReader reader;
        private StringTokenizer tokenizer;

        InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream));
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                tokenizer = new StringTokenizer(readLine());
            }
            return tokenizer.nextToken();
        }

        String readLine() {
            String line;
            try {
                line = reader.readLine();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
            return line;
        }
    }

    static class OutputWriter {
        private final PrintWriter writer;

        OutputWriter(OutputStream stream) {
            writer = new PrintWriter(stream);
        }

        void println(int[] arr) {
            for (int i = 0; i < arr.length; i++) {
                writer.print(arr[i]);
                writer.print(' ');
            }
            writer.println();
        }

        void close() {
            writer.close();
        }
    }

    static class Pair<T, U> {
        T first;
        U second;

        Pair(T first, U second) {
            this.first = first;
            this.second = second;
        }
    }
}

File: output/xCodeEval_remove_8000/C++/Python/fa5bbd62e38df804504d474d91851ba6.py
import sys
from math import atan2, pi

class CPaintTree:
    class Solver:
        def __init__(self, in_stream, out_stream):
            self.in_stream = in_stream
            self.out_stream = out_stream

        def read_tree(self):
            self.e = []
            self.n = int(self.in_stream.readline())
            for _ in range(self.n - 1):
                u, v = map(int, self.in_stream.readline().split())
                self.e.append((u - 1, v - 1))

        def read_points(self):
            self.xy = []
            for _ in range(self.n):
                x, y = map(int, self.in_stream.readline().split())
                self.xy.append((x, y))

        def get_angle(self, i, j):
            return atan2(self.xy[j][1] - self.xy[i][1], self.xy[j][0] - self.xy[i][0])

        def sort_by_angle(self, l, r, cur_xy):
            if r < l:
                return
            max_angle = self.get_angle(cur_xy, self.xys_idx[l][1])
            for i in range(l, r + 1):
                angle = self.get_angle(cur_xy, self.xys_idx[i][1])
                if angle + pi < max_angle:
                    angle += 2 * pi
                self.xys_idx[i] = (angle, self.xys_idx[i][1])
            self.xys_idx[l:r + 1] = sorted(self.xys_idx[l:r + 1])

        def solve_rec(self, l, r, root, parent):
            curr_xy = self.xys_idx[r][1]
            self.xy_mapping[curr_xy] = root
            self.sort_by_angle(l, r, curr_xy)
            cur_l = l
            for i in range(len(self.e[root])):
                j = self.e[root][i]
                if j == parent:
                    continue
                cur_r = cur_l + self.subtree_sz[j] - 1
                self.solve_rec(cur_l, cur_r, j, root)
                cur_l = cur_r + 1

        def calculate_subtree_sizes(self, i, p):
            self.subtree_sz[i] = 1
            for j in self.e[i]:
                if j == p:
                    continue
                self.calculate_subtree_sizes(j, i)
                self.subtree_sz[i] += self.subtree_sz[j]

        def solve(self):
            self.read_tree()
            self.read_points()
            self.xy_mapping = [0] * self.n
            self.subtree_sz = [0] * self.n
            self.calculate_subtree_sizes(0, -1)
            self.xys_idx = [(0, i) for i in range(self.n)]
            self.xys_idx.sort(key=lambda x: (self.xy[x[1]][0], self.xy[x[1]][1]), reverse=True)
            self.solve_rec(0, self.n - 1, 0, -1)
            self.xy_mapping = [x + 1 for x in self.xy_mapping]
            self.out_stream.write(' '.join(map(str, self.xy_mapping)) + '\n')

    @staticmethod
    def solve(in_stream, out_stream):
        CPaintTree.Solver(in_stream, out_stream).solve()

if __name__ == '__main__':
    CPaintTree.solve(sys.stdin, sys.stdout)
